### PR TITLE
PIM-5593: Keep context in the associations tab of the product edit form

### PR DESCRIPTION
**Pull request by phaseinducer**
**Description (for Contributor and Core Developer)**

The state of the following elements is now memorized in session storage (idem products grid) :

associations

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | n/a
| Added Behats                      | yes
| Changelog updated                 | yes
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) | 
| Migration script                  | n/a
| Tech Doc                          | n/a

### CODE REVIEW

File: features/Context/WebUser.php [hunk 1]
```diff
@@ -248,8 +248,35 @@ public function iClickOnTheACLRole($group)
...
+    }
+
+    /**
+     * @param string $association
+     *
+     * @throws ExpectationException
+     *
+     * @Given /^I should be on the "([^"]*)" association$/
+     */
+    public function iShouldBeOnTheAssociation($association)
```
Comment (juliensnz) [inline features/Context/WebUser.php]: move in a proper context ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 2]
```diff
@@ -42,7 +46,7 @@ define(
             className: 'tab-pane active',
             id: 'product-associations',
             events: {
-                'click .nav-tabs li':                'changeAssociationType',
+                'click #associations-list li':       'changeAssociationType',
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: should be a class

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 3]
```diff
@@ -42,7 +46,7 @@ define(
...
                 'click #association-buttons button': 'changeAssociationTargets'
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: same here

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/product/tab/associations.html [hunk 4]
```diff
@@ -9,8 +9,8 @@
         <div class="sidebar">
             <ul class="nav nav-tabs tab-groups" id="associations-list">
                 <% _.each(associationTypes, function (associationType) { %>
-                    <li data-association-type="<%- associationType.code %>"<% if (associationType.code === state.currentAssociationType) { %> class="active"<% } %>>
-                        <a href="javascript:void(0);">
+                    <li data-association-type="<%- associationType.code %>"<% if (associationType.code === currentAssociationType) { %> class="active"<% } %>>
+                        <a href="#">
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/product/tab/associations.html]: I fear `#` in link :(

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 5]
```diff
@@ -173,6 +179,18 @@ define(
                     this.render();
                 }
             },
+            setCurrentAssociationType: function (associationType) {
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: jsdoc

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 6]
```diff
@@ -173,6 +179,18 @@ define(
...
+                sessionStorage.setItem('current_association_type', associationType);
+            },
+            getCurrentAssociationType: function () {
+                return sessionStorage.getItem('current_association_type');
+            },
+            setCurrentAssociationTarget: function (associationType) {
+                sessionStorage.setItem('current_association_target', associationType);
+            },
+            getCurrentAssociationTarget: function () {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: jsdoc for the 4 methods please :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 7]
```diff
@@ -187,49 +205,78 @@ define(
...
+                    collection.processFiltersParams(urlParams, filters, gridName + '[_filter]');
+                }
+
                 $.get(Routing.generate('pim_datagrid_load', urlParams)).then(function (resp) {
-                    $('#grid-' + alias).data({ 'metadata': resp.metadata, 'data': JSON.parse(resp.data) });
+                    this.$('#grid-' + gridName).data({ 'metadata': resp.metadata, 'data': JSON.parse(resp.data) });
 
-                    require(resp.metadata.requireJSModules, function () {
+                    var gridModules = resp.metadata.requireJSModules;
+                    gridModules.push('pim/datagrid/state-listener');
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: I don't understand it, maybe you can explain it irl?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 8]
```diff
@@ -187,49 +205,78 @@ define(
...
             changeAssociationTargets: function (event) {
                 var associationTarget = event.currentTarget.dataset.associationTarget;
-                this.state.set('associationTarget', associationTarget);
+                this.setCurrentAssociationTarget(associationTarget);
 
                 _.each(this.datagrids, function (datagrid, gridType) {
-                    $('#' + datagrid.name)[gridType === associationTarget ? 'removeClass' : 'addClass']('hide');
-                });
+                    var method = gridType === associationTarget ? 'removeClass' : 'addClass';
+                    this.$('#' + datagrid.name)[method]('hide');
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: You can use classes

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 9]
```diff
@@ -187,49 +205,78 @@ define(
...
+                }.bind(this));
+
+                $(event.currentTarget)
+                    .addClass('hide')
+                    .siblings('[data-association-target]')
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: maybe add a class in the HTML ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 10]
```diff
@@ -187,49 +205,78 @@ define(
...
+                if (null !== datagridState.filters) {
+                    var collection = new PageableCollection();
+                    var filters    = collection.decodeStateData(datagridState.filters);
+
+                    collection.processFiltersParams(urlParams, filters, gridName + '[_filter]');
+                }
+
                 $.get(Routing.generate('pim_datagrid_load', urlParams)).then(function (resp) {
-                    $('#grid-' + alias).data({ 'metadata': resp.metadata, 'data': JSON.parse(resp.data) });
+                    this.$('#grid-' + gridName).data({ 'metadata': resp.metadata, 'data': JSON.parse(resp.data) });
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: same here

Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: same here

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 11]
```diff
@@ -346,6 +393,19 @@ define(
...
 
             /**
+             * Return if the specified grid is already rendered
+             *
+             * @param {Object} grid
+             *
+             * @returns {boolean}
+             */
+            isGridRendered: function (grid) {
+                return 0 < this.$('#grid-' + grid.name)
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: :vomit:

**Comment by grena** [comment]
Some questions & comments, then GTM =)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js (hunk 11 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: :dancer:

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/product/tab/associations.html (hunk 4 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/product/tab/associations.html]: No fear : https://github.com/akeneo/pim-community-dev/pull/4210/files#diff-d9ddc47bfa132c48f10e5ecf788b42b8R229

**Comment by nidup** [comment]
GTM

**Pull request merged**